File: 0343-integer-break/0343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        dp = [0] * (n + 1)
        dp[1] = 1

        for i in range(2, n + 1):
            max_product = 0
            for j in range(1, i):
                max_product = max(
                    max_product,
                    max(j * (i - j), j * dp[i - j])
                )
            dp[i] = max_product

        return dp[n]

File: 0343-integer-break/0343-integer-break.java
class Solution {
    public int integerBreak(int n) {
        int[] dp = new int[n + 1];
        dp[1] = 1;

        for (int i = 2; i <= n; i++) {
            int maxProduct = 0;
            for (int j = 1; j < i; j++) {
                maxProduct = Math.max(maxProduct, 
                                Math.max(j * (i - j), j * dp[i - j]));
            }
            dp[i] = maxProduct;
        }
        return dp[n];    
    }
}